File: android/app/src/main/java/app/releaf/mobile/data/notebook/ChapterEntity.kt
/*
 * ChapterEntity.kt
 *
 * Room entity mirroring the `chapters` table in
 * `design-system/migrations/v1_initial.sql`. Chapters are a middle tier in
 * the notebook → chapter → page hierarchy: each chapter belongs to exactly
 * one notebook and groups a linear sequence of pages.
 *
 * FK to notebooks is declared in the .sql file; Room's @ForeignKey is
 * intentionally omitted for parity with NotepadEntry (see that file for the
 * schema-source-of-truth follow-up). Cascade semantics are handled at the
 * repository layer today.
 */

package app.releaf.mobile.data.notebook

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "chapters",
    indices = [
        // Composite supports the hot query "chapters of a notebook, in order".
        Index(value = ["notebook_id", "position"]),
        Index("updated_at"),
        Index("deleted_at"),
        Index("archived_at"),
    ],
)
data class ChapterEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "notebook_id")
    val notebookId: String,

    @ColumnInfo(name = "title")
    val title: String,

    /** Free-form summary shown on the chapter row + chapter-detail card. */
    @ColumnInfo(name = "description")
    val description: String? = null,

    @ColumnInfo(name = "position", defaultValue = "1024")
    val position: Long = 1024L,

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,

    /**
     * Soft-archive timestamp (separate from `deletedAt`). Added in
     * v17 — non-nil means the chapter lives in the archive bin and
     * is filtered out of the active chapter list. Independent of
     * `deletedAt` (true tombstone).
     */
    @ColumnInfo(name = "archived_at")
    val archivedAt: String? = null,
)

File: apps/releaf/ios/Releaf/Data/Notebook/ChapterEntity.swift
/*
 * ChapterEntity.swift
 *
 * GRDB record for the `chapters` table. Chapters are the middle tier
 * in the notebook → chapter → page hierarchy — each belongs to exactly
 * one notebook and groups a linear sequence of pages.
 *
 * Like `NotebookEntity`, the unsuffixed `Chapter` type lives in
 * `Data/Domain/Models.swift` as the Drive-facing shape; this struct is
 * the persistence shape.
 */

import Foundation
import GRDB

public struct ChapterEntity: Codable, FetchableRecord, PersistableRecord,
                             Identifiable, Equatable, Sendable {

    public static let databaseTableName = "chapters"

    public var id: String
    public var notebookId: String
    public var title: String
    public var position: Int64
    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?
    /// Soft-archive timestamp (separate from `deletedAt`). Added in
    /// v6 — non-nil means the chapter lives in the archive bin and
    /// is filtered out of the active chapter list.
    public var archivedAt: String?

    public enum CodingKeys: String, CodingKey {
        case id
        case notebookId   = "notebook_id"
        case title
        case position
        case driveFileId  = "drive_file_id"
        case createdAt    = "created_at"
        case updatedAt    = "updated_at"
        case dirty
        case deletedAt    = "deleted_at"
        case archivedAt   = "archived_at"
    }

    public init(
        id: String,
        notebookId: String,
        title: String,
        position: Int64 = 1024,
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil,
        archivedAt: String? = nil
    ) {
        self.id = id
        self.notebookId = notebookId
        self.title = title
        self.position = position
        self.driveFileId = driveFileId
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.dirty = dirty
        self.deletedAt = deletedAt
        self.archivedAt = archivedAt
    }
}
